include/cuco/detail/priority_queue.inl

```diff
@@ -0,0 +1,178 @@
+#pragma once
+#include <cmath>
+
+#include <cuco/detail/error.hpp>
+#include <cuco/detail/priority_queue_kernels.cuh>
+
+namespace cuco {
+
+template <typename T, typename Compare, bool FavorInsertionPerformance, typename Allocator>
+priority_queue<T, Compare, FavorInsertionPerformance, Allocator>::priority_queue(
+  size_t initial_capacity, Allocator const& allocator)
+  : allocator_{allocator},
+    int_allocator_{allocator},
+    t_allocator_{allocator},
+    size_t_allocator_{allocator}
+{
+  node_size_ = NodeSize;
+
+  // Round up to the nearest multiple of node size
+  int nodes = ((initial_capacity + node_size_ - 1) / node_size_);
+
+  node_capacity_      = nodes;
+  lowest_level_start_ = 1 << (int)log2(nodes);
+
+  // Allocate device variables
+
+  d_size_ = std::allocator_traits<int_allocator_type>::allocate(int_allocator_, 1);
+
+  CUCO_CUDA_TRY(cudaMemset(d_size_, 0, sizeof(int)));
+
+  d_p_buffer_size_ = std::allocator_traits<size_t_allocator_type>::allocate(size_t_allocator_, 1);
+
+  CUCO_CUDA_TRY(cudaMemset(d_p_buffer_size_, 0, sizeof(size_t)));
+
+  d_heap_ = std::allocator_traits<t_allocator_type>::allocate(
+    t_allocator_, node_capacity_ * node_size_ + node_size_);
+
+  d_locks_ =
+    std::allocator_traits<int_allocator_type>::allocate(int_allocator_, node_capacity_ + 1);
+
+  CUCO_CUDA_TRY(cudaMemset(d_locks_, 0, sizeof(int) * (node_capacity_ + 1)));
+}
+
+template <typename T, typename Compare, bool FavorInsertionPerformance, typename Allocator>
+priority_queue<T, Compare, FavorInsertionPerformance, Allocator>::~priority_queue()
+{
+  std::allocator_traits<int_allocator_type>::deallocate(int_allocator_, d_size_, 1);
+  std::allocator_traits<size_t_allocator_type>::deallocate(size_t_allocator_, d_p_buffer_size_, 1);
+  std::allocator_traits<t_allocator_type>::deallocate(
+    t_allocator_, d_heap_, node_capacity_ * node_size_ + node_size_);
+  std::allocator_traits<int_allocator_type>::deallocate(
+    int_allocator_, d_locks_, node_capacity_ + 1);
+}
+
+template <typename T, typename Compare, bool FavorInsertionPerformance, typename Allocator>
+template <typename InputIt>
+void priority_queue<T, Compare, FavorInsertionPerformance, Allocator>::push(InputIt first,
+                                                                            InputIt last,
+                                                                            cudaStream_t stream)
+{
+  const int kBlockSize = min(256, (int)node_size_);
```

Comment:
```suggestion
  const int kBlockSize = std::min(256, static<int>(node_size_));
```
Please avoid c-style expressions.

Actually, there is an issue here: what if node size is not a multiple of warp size and is smaller than 256? For the performance purpose, the block size should always be a multiple of 32. The common choices are 128 and 256. We may want to determine grid/block dim like other cuco implementations (see [here](https://github.com/NVIDIA/cuCollections/blob/edb0bfda0322b3595e56db9db05427401bed732e/include/cuco/detail/static_map.inl#L134-L137))?

Reply:
Currently, node size is determined [here](https://github.com/andrewbriand/cuCollections/blob/dev/include/cuco/priority_queue.cuh#L77) according to the template parameter `FavorInsertionPerformance`, and so is always a multiple of 32. Block size depends on node size because any extra threads past the node size in each block will simply sit idle in the internal heap's operations.

I see that in other cuco implementations, block size is usually fixed. However, as discussed briefly with Jake [above](https://github.com/NVIDIA/cuCollections/pull/105#discussion_r848905152), I included `FavorInsertionPerformance` because I was thinking that there might be some applications where the user inserts a very large amount of data to the queue but is only interested in retrieving the top few elements. I now think though that removing this may be better, given that the performance increase to insertion is not that significant (4.3s to 3.8s on a Tesla K80 for 256M elements in my testing), and that it would simplify the code.

Reply:
Ok, I've removed `FavorInsertionPerformance` in 55cf2e6